src/frontend/src/expr/type_inference/func.rs

```diff
@@ -200,40 +234,114 @@ fn infer_type_for_special(
+            let common_ele_type = common_ele_type.unwrap();
+            let most_nested = get_most_nested(left_type.clone(), right_type.clone());
+            let array_type = add_nesting(common_ele_type.clone(), most_nested);
+
+            // try to cast inputs to inputs to common type
+            let inputs_owned = std::mem::take(inputs);
+            let try_cast = inputs_owned
+                .into_iter()
+                .map(|input| {
+                    let x = input.return_type();
+                    input.cast_explicit(add_nesting(common_ele_type.clone(), x))
+                })
+                .try_collect();
+
+            match try_cast {
+                Ok(casted) => {
+                    // apply type conversion and return common type
+                    *inputs = casted;
+                    Ok(Some(array_type))
+                }
+                Err(_) => Err(ErrorCode::BindError(format!(
+                    "Cannot append {} to {}",
+                    right_type, left_type
+                ))
+                .into()),
+            }
         }
         ExprType::ArrayPrepend => {
             ensure_arity!("array_prepend", | inputs | == 2);
+            // get input types
             let left_type = inputs[0].return_type();
             let right_type = inputs[1].return_type();
-            let return_type = match (&left_type, &right_type) {
-                (DataType::List { .. }, DataType::List { .. }) => None,
-                (
-                    _, // non-array
-                    DataType::List {
-                        datatype: right_elem_type,
-                    },
-                ) if left_type == **right_elem_type => Some(right_type.clone()),
+            let right_ele_type_opt = match &right_type {
+                DataType::List { datatype: right_et } => Some(right_et),
                 _ => None,
             };
-            Ok(Some(return_type.ok_or_else(|| {
-                ErrorCode::BindError(format!("Cannot prepend {} to {}", left_type, right_type))
-            })?))
+            let right_ele_type = right_ele_type_opt.ok_or_else(|| {
+                ErrorCode::BindError(format!("Cannot prepend {} to {}", right_type, left_type))
+            })?;
+
+            // cast to least restrictive type or return error
+            let common_ele_type = least_restrictive(*right_ele_type.clone(), left_type.clone());
+            let nesting_level_diff = (calc_nesting_level(left_type.clone())
+                - calc_nesting_level(right_type.clone()))
+            .abs();
+            if common_ele_type.is_err() || nesting_level_diff > 1 {
+                return Err(ErrorCode::BindError(format!(
+                    "unable to find least restrictive type between {} and {}",
+                    left_type, right_type
+                ))
+                .into());
+            }
+
+            // found common type
+            let common_ele_type = common_ele_type.unwrap();
```

Comment:
`nesting_level_diff` check seems useless here. Maybe you are just trying to follow `array_cat` so that they can share a common utility function later. I will share my idea of this utility in the final comment.

Without `nesting_level_diff`, it could be just:
```
             let common_ele_type = least_restrictive(*right_ele_type.clone(), left_type.clone())
                 .map_err(|_| {
                     ErrorCode::BindError(format!(
                         "unable to find least restrictive type between {} and {}",
                         left_type, right_type
                     ))
                 })?;
```

Reply:
> I am not sure that `nesting_level_diff` is useless. If you want to combine `array[array[]::int]` with an `int` you need the `nesting_level_diff` to detect that this is not a valid append/prepend/concat, even though a `common_ele_type` is found

For context, my "useless" claim only applies to append/prepend but not concat. Assume we attempt to prepend `int` to `array[array[]::int]`. The `least_restrictive(*right_ele_type.clone(), left_type.clone())` check would already report an error because there is least restrictive between `int[]` (`right_ele_type`) and `int` (`left_type`).

Reply:
Agreed. I think using just retrying `align_array_and_element` in concat with a different input is a good solution that saves us the extra work of the `nested_level_diff` helper functions.  

```rust 
let common_type = align_array_and_element(0, 1, inputs).or_else(|_| align_array_and_element(1, 0, inputs));
```